### PR TITLE
Fix getting control instance

### LINKED ISSUE #306: [2pt] Dead instance can be selected to be one not expelled

### PR DESCRIPTION
**Pull request by dokshina**
* Now one not expelled instance should also be alive - it's checked
  by creating a connection using instances advertise URIs
* Control instance should be alive
* If there are some joined instances, but no one of them isn't alive -
  getting control instance fails.

Closes #306

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,12 @@ README.md to use the newest tag with new release
 - Missing default config for machine with stateboard
 - Specifying `cartridge_app_name` other than the TGZ package name
 - Creating unnamed replicasets with instances without `replicaset_alias` set
+- Getting control instance:
+  - Now one not expelled instance should also be alive; it's checked
+    by creating connection using instances advertise URIs
+  - Control instance should be alive
+  - I there are some joined instances, but no one of them isn't alive,
```
Comment (hackallcode) [inline CHANGELOG.md]:
```suggestion
  - If there are some joined instances, but no one of them isn't alive,
```

File: library/cartridge_get_alive_not_expelled_instance.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    app_name = params['app_name']
+
+    canditates_by_uris = {}
+
+    for instance_name in play_hosts:
+        instance_vars = module_hostvars[instance_name]
+        if helpers.is_expelled(instance_vars) or helpers.is_stateboard(instance_vars):
+            continue
+
+        advertise_uri = instance_vars['config']['advertise_uri']
```
Comment (hackallcode) [inline library/cartridge_get_alive_not_expelled_instance.py]: What about to check that `config` and `advertise_uri` sections exists?

File: library/cartridge_get_alive_not_expelled_instance.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        canditates_by_uris[advertise_uri] = instance_name
+
+    alive_not_expelled_instance_name = None
+    for uri, instance_name in sorted(canditates_by_uris.items()):
+        host, port = uri.split(':', 1)
```
Comment (hackallcode) [inline library/cartridge_get_alive_not_expelled_instance.py]:
```suggestion
        host, port = uri.rsplit(':', 1)
```

File: library/cartridge_get_alive_not_expelled_instance.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        try:
+            conn = socket.create_connection((host, port))
+            conn.recv(1024)
```
Comment (hackallcode) [inline library/cartridge_get_alive_not_expelled_instance.py:44]: What are you getting the data for? Can we remove it? Can we reduce size of buffer?

File: library/cartridge_get_control_instance.py [hunk 5]
```diff
@@ -85,32 +83,51 @@ def get_control_instance(params):
         return require('membership').members()
     ''')
 
+    alive_members = set()
+    found_joined_instances = False
+
     for uri, member in sorted(members.items()):
         if 'payload' not in member or not member['payload']:
```
Comment (hackallcode) [inline library/cartridge_get_control_instance.py]:
```suggestion
        if not member.get('payload'):
```

File: library/cartridge_get_control_instance.py [hunk 6]
```diff
@@ -85,32 +83,51 @@ def get_control_instance(params):
...
             return helpers.ModuleRes(failed=True, msg='Instance %s does not contain payload' % uri)
 
-        if member.get('status') != 'alive':
+        member_payload = member['payload']
+        if member_payload.get('alias') is None:
+            return helpers.ModuleRes(failed=True, msg='Instance %s payload does not contain alias' % uri)
+
+        instance_name = member_payload['alias']
```
Comment (hackallcode) [inline library/cartridge_get_control_instance.py]:
```suggestion
        instance_name = member_payload.get('alias')
        if instance_name is None:
            return helpers.ModuleRes(failed=True, msg='Instance %s payload does not contain alias' % uri)

```

File: library/cartridge_get_control_instance.py [hunk 7]
```diff
@@ -85,32 +83,51 @@ def get_control_instance(params):
...
+
+        if member_payload.get('uuid') is None:
             continue
 
-        member_payload = member['payload']
-        if member_payload.get('uuid') is not None:
-            if member_payload.get('alias') is None:
-                return helpers.ModuleRes(failed=True, msg='Instance %s payload does not contain alias' % uri)
+        if instance_name not in module_hostvars:
+            continue
```
Comment (hackallcode) [inline library/cartridge_get_control_instance.py:108]: in play_hosts?

File: library/cartridge_get_alive_not_expelled_instance.py (hunk 2 above)
Comment (dokshina) [reply library/cartridge_get_alive_not_expelled_instance.py]: We've checked it on the `validate_config` step. Now it's an internal error.

**Review by hackallcode — DISMISSED** [review]

**Review by hackallcode — APPROVED** [review]
Good patch!

**Pull request merged**